File: frontend/src/lib/weeklyGoal.ts
import { isoWeekNumber, mondayOf, weekdayMonFirst, ymd } from './dates';
import type { Task, TargetSegment } from './types';

export type SegmentProgress = {
  segment: TargetSegment;
  done: number;
  hit: boolean;
  exceeded: boolean;
};

export type GoalStatus = {
  done: number;
  target: number; // total target (sum of segments when present, else target_per_week)
  hit: boolean; // every segment (or flat goal) reached its target
  exceeded: boolean; // any segment exceeded, or flat done > target
  segments: SegmentProgress[]; // empty when the task is a flat target_per_week goal
};

function inSameIsoWeek(a: Date, b: Date): boolean {
  return a.getFullYear() === b.getFullYear() && isoWeekNumber(a) === isoWeekNumber(b);
}

/** Count completions in the ISO week containing [refDate], optionally filtered by weekdays. */
function countInWeek(t: Task, refDate: Date, weekdays?: number[]): number {
  return t.completed_dates.reduce((acc, cd) => {
    const d = new Date(cd + 'T00:00:00');
    if (!inSameIsoWeek(d, refDate)) return acc;
    if (weekdays && !weekdays.includes(weekdayMonFirst(d))) return acc;
    return acc + 1;
  }, 0);
}

export function goalStatus(t: Task, refDateYMD: string): GoalStatus {
  const ref = new Date(refDateYMD + 'T00:00:00');
  const segments = t.target_segments ?? [];
  if (segments.length === 0) {
    const target = t.target_per_week ?? 0;
    const done = countInWeek(t, ref);
    return {
      done,
      target,
      hit: target > 0 && done >= target,
      exceeded: target > 0 && done > target,
      segments: [],
    };
  }
  const perSegment: SegmentProgress[] = segments.map((seg) => {
    const done = countInWeek(t, ref, seg.weekdays);
    return {
      segment: seg,
      done,
      hit: done >= seg.target,
      exceeded: done > seg.target,
    };
  });
  const done = perSegment.reduce((a, s) => a + s.done, 0);
  const target = segments.reduce((a, s) => a + s.target, 0);
  return {
    done,
    target,
    hit: perSegment.every((p) => p.hit),
    exceeded: perSegment.some((p) => p.exceeded),
    segments: perSegment,
  };
}

/**
 * Cumulative count of completions in the segment that owns [dateYMD]'s
 * weekday, counting every completion in the ISO week with date <= dateYMD.
 * Used to show "(N/T)" in the row title. Handles duplicate completions
 * (same date twice) correctly because they're counted separately.
 */
export function weeklyGoalRank(t: Task, dateYMD: string): number {
  const d = new Date(dateYMD + 'T00:00:00');
  const wd = weekdayMonFirst(d);
  const segs = t.target_segments ?? [];
  const seg = segs.find((s) => s.weekdays.includes(wd));
  const filter = seg ? seg.weekdays : undefined;
  return t.completed_dates.reduce((acc, cd) => {
    if (cd > dateYMD) return acc;
    const cdd = new Date(cd + 'T00:00:00');
    if (!inSameIsoWeek(cdd, d)) return acc;
    if (filter && !filter.includes(weekdayMonFirst(cdd))) return acc;
    return acc + 1;
  }, 0);
}

export function weeklyGoalLabel(t: Task, dateYMD: string): string {
  const d = new Date(dateYMD + 'T00:00:00');
  const wd = weekdayMonFirst(d);
  const segs = t.target_segments ?? [];
  const seg = segs.find((s) => s.weekdays.includes(wd));
  const k = weeklyGoalRank(t, dateYMD);
  const tgt = seg?.target ?? t.target_per_week ?? '?';
  return `${t.title} (${k}/${tgt})`;
}

export type StreakInfo = {
  current: number; // weeks running up to (and including) the current week if hit
  best: number; // longest historical run
};

/**
 * Consecutive ISO weeks where the goal hit its target. The current week
 * counts only if it is already hit; otherwise the streak starts from the
 * previous week. Weeks before the task was created are not counted.
 */
export function streakInfo(t: Task, todayYMD: string): StreakInfo {
  const today = new Date(todayYMD + 'T00:00:00');
  const startSource = t.start_date ? new Date(t.start_date + 'T00:00:00') : new Date(t.created_at);
  const startMonday = mondayOf(startSource);
  // Cap scan at end_date when set; otherwise today.
  const endSource = t.end_date ? new Date(t.end_date + 'T00:00:00') : today;
  const endMonday = mondayOf(endSource < today ? endSource : today);
  let cursor = endMonday;

  const hit = (d: Date) => goalStatus(t, ymd(d)).hit;

  let current = 0;
  if (!hit(cursor)) {
    cursor.setDate(cursor.getDate() - 7);
  }
  while (cursor >= startMonday && hit(cursor)) {
    current++;
    cursor.setDate(cursor.getDate() - 7);
  }

  // Best streak: scan every week in the active range.
  let best = 0;
  let run = 0;
  let scan = new Date(startMonday);
  while (scan <= endMonday) {
    if (hit(scan)) {
      run++;
      if (run > best) best = run;
    } else {
      run = 0;
    }
    scan.setDate(scan.getDate() + 7);
  }
  return { current, best };
}

export const WEEKDAY_SHORT = ['Mon', 'Tue', 'Wed', 'Thu', 'Fri', 'Sat', 'Sun'] as const;

export function segmentLabel(seg: TargetSegment): string {
  // Compact label e.g. "Mon-Thu" when contiguous, else "Mon · Wed · Fri".
  const sorted = [...seg.weekdays].sort((a, b) => a - b);
  if (sorted.length === 0) return '';
  if (sorted.length === 1) return WEEKDAY_SHORT[sorted[0]];
  const contiguous = sorted.every((v, i) => i === 0 || v === sorted[i - 1] + 1);
  if (contiguous) return `${WEEKDAY_SHORT[sorted[0]]}-${WEEKDAY_SHORT[sorted[sorted.length - 1]]}`;
  return sorted.map((w) => WEEKDAY_SHORT[w]).join(' · ');
}

File: android/app/src/main/java/com/joncas/jontracker/data/WeeklyGoal.kt
package com.joncas.jontracker.data

import com.joncas.jontracker.api.Task
import com.joncas.jontracker.api.TargetSegment
import java.time.LocalDate
import java.time.temporal.WeekFields

data class SegmentProgress(
    val segment: TargetSegment,
    val done: Int,
    val hit: Boolean,
    val exceeded: Boolean,
)

data class GoalStatus(
    val done: Int,
    val target: Int,
    val hit: Boolean,
    val exceeded: Boolean,
    val segments: List<SegmentProgress>,
)

private fun Task.completionsInWeek(refDate: LocalDate, weekdays: List<Int>? = null): Int {
    val refWeek = refDate.get(WeekFields.ISO.weekOfWeekBasedYear())
    val refYear = refDate.year
    return completed_dates.count { d ->
        val cd = runCatching { LocalDate.parse(d) }.getOrNull() ?: return@count false
        if (cd.year != refYear || cd.get(WeekFields.ISO.weekOfWeekBasedYear()) != refWeek) return@count false
        if (weekdays == null) true else weekdays.contains(cd.weekdayMonFirst())
    }
}

fun Task.goalStatus(refDate: LocalDate): GoalStatus {
    val segs = target_segments ?: emptyList()
    if (segs.isEmpty()) {
        val target = target_per_week ?: 0
        val done = completionsInWeek(refDate)
        return GoalStatus(
            done = done,
            target = target,
            hit = target > 0 && done >= target,
            exceeded = target > 0 && done > target,
            segments = emptyList(),
        )
    }
    val perSegment = segs.map { seg ->
        val done = completionsInWeek(refDate, seg.weekdays)
        SegmentProgress(
            segment = seg,
            done = done,
            hit = done >= seg.target,
            exceeded = done > seg.target,
        )
    }
    return GoalStatus(
        done = perSegment.sumOf { it.done },
        target = segs.sumOf { it.target },
        hit = perSegment.all { it.hit },
        exceeded = perSegment.any { it.exceeded },
        segments = perSegment,
    )
}

/**
 * Cumulative number of completions in the segment that owns [date]'s weekday,
 * counting every completion in the ISO week with date <= [date]. Handles
 * duplicate completions for the same day (weekly goals can be ticked many
 * times per day) by counting each separately.
 */
fun Task.weeklyGoalRank(date: LocalDate): Int {
    val wd = date.weekdayMonFirst()
    val filter = (target_segments ?: emptyList()).firstOrNull { it.weekdays.contains(wd) }?.weekdays
    val refWeek = date.get(WeekFields.ISO.weekOfWeekBasedYear())
    val refYear = date.year
    return completed_dates.count { d ->
        val cd = runCatching { LocalDate.parse(d) }.getOrNull() ?: return@count false
        if (cd.year != refYear || cd.get(WeekFields.ISO.weekOfWeekBasedYear()) != refWeek) return@count false
        if (cd.isAfter(date)) return@count false
        if (filter == null) true else filter.contains(cd.weekdayMonFirst())
    }
}

fun Task.displayTitle(date: LocalDate): String {
    if (task_type != "weekly_goal") return title
    val wd = date.weekdayMonFirst()
    val seg = (target_segments ?: emptyList()).firstOrNull { it.weekdays.contains(wd) }
    val target = seg?.target ?: target_per_week ?: 0
    val rank = weeklyGoalRank(date).coerceAtLeast(0)
    val shown = if (rank == 0) completionsInWeek(date, seg?.weekdays) else rank
    return "$title ($shown/$target)"
}

data class StreakInfo(val current: Int, val best: Int)

/**
 * Consecutive ISO weeks where the goal hit its target. Current week counts
 * only if already hit; otherwise the streak starts from the previous week.
 * Weeks before [Task.created_at] aren't counted.
 */
fun Task.streakInfo(today: LocalDate): StreakInfo {
    val createdMonday = runCatching {
        LocalDate.parse(created_at.substring(0, 10))
    }.getOrNull()?.mondayOfWeek() ?: today.mondayOfWeek()
    val endMonday = today.mondayOfWeek()

    var cursor = endMonday
    if (!goalStatus(cursor).hit) cursor = cursor.minusWeeks(1)
    var current = 0
    while (!cursor.isBefore(createdMonday) && goalStatus(cursor).hit) {
        current++
        cursor = cursor.minusWeeks(1)
    }

    var best = 0
    var run = 0
    var scan = createdMonday
    while (!scan.isAfter(endMonday)) {
        if (goalStatus(scan).hit) {
            run++
            if (run > best) best = run
        } else {
            run = 0
        }
        scan = scan.plusWeeks(1)
    }
    return StreakInfo(current, best)
}

private val WEEKDAY_SHORT = listOf("Mon", "Tue", "Wed", "Thu", "Fri", "Sat", "Sun")

fun segmentLabel(seg: TargetSegment): String {
    val sorted = seg.weekdays.sorted()
    if (sorted.isEmpty()) return ""
    if (sorted.size == 1) return WEEKDAY_SHORT[sorted[0]]
    val contiguous = sorted.zipWithNext().all { (a, b) -> b == a + 1 }
    return if (contiguous) "${WEEKDAY_SHORT[sorted.first()]}-${WEEKDAY_SHORT[sorted.last()]}"
    else sorted.joinToString(" · ") { WEEKDAY_SHORT[it] }
}
